### PR TITLE
test command improvments

### PR DESCRIPTION
**Pull request by PierreJeanjacquot**
- limit test memory to 1Go (worker condition) and detect docker out-of-memory errors
image

- report container exited with error code (!= 0)
image

- stream test logs, collapse them, and propose to display them after run
image
image

- stream docker build logs and collapse when successful
image
image
image

- display output
image
image

Misc
- always run with docker (to better reflect the worker environment)
- refactor docker module (remove logging responsibility)
- start refactoring logging with ora spinner (WIP)
  - use only one spinner instance and log from top level
  - always start the spinner explaining what is going on
  - stop the spinner for fatal errors and display error
  - persist useful info only with correct level

### CODE REVIEW

File: cli/src/test.js [hunk 1]
```diff
@@ -10,100 +7,119 @@ import {
...
 
-  const { withProtectedData } = await readIDappConfig(spinner);
-  spinner.succeed('Reading idapp JSON config file.');
+  // just start the spinner, no need to persist success in terminal
+  spinner.start('Checking docker daemon is running...');
+  await checkDockerDaemon();
+  // build a temp image for test
+  spinner.start('Building app docker image for test...\n');
+  const imageId = await dockerBuild({
+    isForTest: true, // Adjust based on your logic
```
Comment (cedric25) [inline cli/src/test.js:67]: We can probably clean this comment

```suggestion
    isForTest: true,
```

File: cli/src/test.js [hunk 2]
```diff
@@ -10,100 +7,119 @@ import {
...
-    await testWithDocker(argv.params);
-  } else {
-    await testWithoutDocker(argv.params);
+  const spinner = ora();
+  try {
+    await cleanTestOutput({ spinner });
+    await testApp({ args: argv.params, spinner });
+    // TODO check output files
+    // - output required files
+    // - outpout dir size
```
Comment (cedric25) [inline cli/src/test.js:24]:
```suggestion
    // - output dir size
```

File: cli/src/test.js [hunk 3]
```diff
@@ -10,100 +7,119 @@ import {
...
+    await askShowTestOutput({ spinner });
```
Comment (cedric25) [inline cli/src/test.js:25]: Should maybe not propose to see the results if app exited with an error?

**Review by cedric25 — CHANGES_REQUESTED** [review]
- Now that option to test outside of docker has disappeared, we can remove 2 outdated lines `idapp test` and `idapp test --params your-name` at the end of "idapp init"

 - I didn't find why but I'm having inconsistent results while running "idapp test" or "idapp test --params Cédric"
Sometimes getting:

Capture d’écran 2024-11-25 à 16 42 10

in the idapp code at line:

```
// Write result to /iexec_out/
await writeFile(`${output}/result.txt`, asciiArtText);
```

File: cli/src/deploy.js [hunk 4]
```diff
@@ -52,44 +34,46 @@ export async function deployForDebug() {
...
+        spinner.text = spinner.text + msg;
+      },
+    });
+    spinner.succeed(`Pushed image ${imageTag} on dockerhub`);
+
+    spinner.start(
+      'Transforming your image into a TEE image and deploying on iExec, this may take a few minutes ...'
+    );
     const { dockerHubUrl } = await sconify({
       mainSpinner: sconifySpinner,
```
Comment (cedric25) [inline cli/src/deploy.js:68]: `sconifySpinner` does not seem to exist anymore.

Maybe:
```suggestion
      mainSpinner: spinner,
```

File: cli/src/index.js [hunk 5]
```diff
@@ -19,37 +19,21 @@ yargs(hideBin(process.argv))
...
-          describe: 'Deploy iDapp for production mode',
-          type: 'boolean',
-          default: false,
-        })
-        .option('debug', {
-          describe: 'Deploy iDapp for debug mode',
-          type: 'boolean',
-          default: false,
-        });
+      return yargs;
```
Comment (cedric25) [inline cli/src/index.js:36]: Maybe simpler syntax:

```
(yargs) => {
  return yargs;
}
```

to:


```
() => {},
```

File: cli/src/test.js (hunk 3 above)
Comment (PierreJeanjacquot) [reply cli/src/test.js:25]: IMO it's good to know what was produced or not

File: cli/src/index.js (hunk 5 above)
Comment (PierreJeanjacquot) [reply cli/src/index.js]: removed yargs builder

**Review by cedric25 — APPROVED** [review]

**Pull request merged**